docs/_includes/Statements.md

```diff
@@ -239,6 +240,242 @@ Note that the form
+ * if instead the RHS of the update-with-failure statement is one or more expressions, the first of these expressions must be a value with a failure-compatible type
+ * if the failure-compatible type of the RHS does not have an Extract method,
+then the LHS of the `:-` statement has one less expression than the RHS 
+(or than the number of out-parameters from the method call)
+ * if the failure-compatible type of the RHS does have an Extract method,
+then the LHS of the `:-` statement has the same numnber of expressions as the RHS 
+(or than the number of out-parameters from the method call)
+and the type of the first LHS expression must be assignable from the return type of the Extract method
+* the IsFailure, PropagateFailure and Extract methods may not be ghost
+
+A failure-compatible type is a type that has the following (each with no in-parameters and one out-parameter):
+
+ * a method or function `IsFailure()` that returns a `bool`
+ * a method or function `PropagateFailure()` that returns a value assignable to the first out-parameter of the caller
+ * an optional method or function `Extract()`
+
+### Failure compatible types
+
+A simple failure-compatible type is the following:
+```dafny
+{% include Example-Fail1.dfy %}
+```
+
+A commonly used alternative that carries some value information is something like this generic type:
+```dafny
+{% include Example-Fail2.dfy %}
+```
+
+A failure-compatible type with an `Extract` method is called _value-carrying_.
+
+
+### Simple status return with no other outputs
+
+The simplest use of this failure-return style of programming is to have a method call that just returns a non-value-carrying `Status` value:
+```dafny
+method Callee(i: int) returns (r: Status)
+{
+  if i < 0 { return Failure("negative"); }
+  return Success;
+}
+
+method Caller(i: int) returns (rr: Status)
+{
+  :- Callee(i);
+  ...
+}
+```
+
+Note that there is no LHS to the `:-` statement. 
+If `Callee` returns `Failure`, then the caller immediately returns, 
+not executing any statements following the call of `Callee`. 
+The value returned by `Caller` (the value of `rr` in the code above) is the result of `PropagateFailure` applied to the value returned by `Callee`, which is often just the same value. 
+If `Callee` does not return `Failure` (that is, returns a value for which `IsFailure()` is `false`) 
+then that return value is forgotten and execution proceeds normally with the statements following the call of `callee` in the body of `Caller`.
+
+### Status return with additional outputs
+
+The example in the previous subsection affects the program only through side effects or the status return itself. 
+It may well be convenient to have additional out-parameters, as is allowed for `:=` updates;
+these out-parameters behave just as for `:=`.
+Here is an example:
+
+```dafny
+method Callee(i: int) returns (r: Status, v: int, w: int)
+{
+  if i < 0 { return Failure("negative"); }
```

Comment:
This `return` statement is not legal. Since the enclosing method has 3 out-parameters, the `return` statement must either have 0 arguments or 3 arguments. Here are three possible ways to make it legal:

* Pass specific out-values for `v` and `w`:
  ```
  return Failure("negative"), 0, 0;
  ```

* Pass the current values of `v` and `w` as out-values:
  ```
  return Failure("negative"), v, w;
  ```

* Only assign to `r`, and then return from the method:
  ```
  r := Failure("negative");
  return;
  ```